### PR TITLE
feature: controller reconciliation max delay

### PR DESCRIPTION
**Pull request by csviri**

### CODE REVIEW

File: docs/documentation/features.md [hunk 1]
```diff
@@ -223,6 +222,27 @@ public class DeploymentReconciler
   }
 ```
 
+## Max Interval Between Reconciliations
+
+In case informers are all in place and reconciler is implemented correctly the reconciliation is triggered when needed,
```
Comment (andreaTP) [inline docs/documentation/features.md:227]: Maybe something like:

```
The reconciliation is supposed to be self-sufficient to cover all the use-cases, however ...
```

**Review by andreaTP — APPROVED** [review]
LGTM

File: docs/documentation/features.md (hunk 1 above)
Comment (csviri) [reply docs/documentation/features.md:227]: will improve the wording, but this does not mean the same, also little bit abstract imho.

Comment (csviri) [reply docs/documentation/features.md:227]: The point is that event sources are needed to have a correct implementation.

File: docs/documentation/features.md [hunk 2]
```diff
@@ -198,8 +196,9 @@ in status) to handle generation filtering in memory. Thus, if an event is receiv
 resource is compared with the resource in the cache.
 
 Note that the **first approach has significant benefits** in the situation when the operator is restarted and there is
-no cached resource anymore. In case two this leads to a reconciliation of every resource, event if the resource is not
-changed while the operator was not running.
+no cached resource anymore. In case two this leads to a reconciliation of every resource in all cases,
```
Comment (metacosm) [inline docs/documentation/features.md:199]: Would have been better to do these changes in a separate PR.

File: docs/documentation/features.md [hunk 3]
```diff
@@ -223,6 +222,28 @@ public class DeploymentReconciler
   }
 ```
 
+## Max Interval Between Reconciliations
+
+In case informers are all in place and reconciler is implemented correctly, there is no need for additional triggers. 
+However, it's a [common practice](https://github.com/java-operator-sdk/java-operator-sdk/issues/848#issuecomment-1016419966) 
+having a failsafe periodic trigger in place,
```
Comment (metacosm) [inline docs/documentation/features.md:229]:
```suggestion
to have a failsafe periodic trigger in place,
```

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java [hunk 4]
```diff
@@ -114,4 +115,12 @@ default boolean useFinalizer() {
   default ResourceEventFilter<R> getEventFilter() {
     return ResourceEventFilters.passthrough();
   }
+
+  default long reconciliationMaxInterval() {
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: I think it will be better to introduce a separate annotation if we split the duration and its unit to make the name simpler and make it more coherent because it doesn't make sense to provide one without the other.

**Review by metacosm — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java (hunk 4 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: I'm a little worried that it would unnecessary complicate things; also there is always a default value. Also both start with same prefix, I don't think this will be confusing.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: An alternative is to have the time as a string, like `"10h"` or "140m". But we use this separation in other API already.

File: docs/documentation/features.md (hunk 2 above)
Comment (csviri) [reply docs/documentation/features.md:199]: it would...will extract it.

Comment (metacosm) [reply docs/documentation/features.md]: No need, just for future reference ;)

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java (hunk 4 above)
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: I don't think it would complicate things and it would make things cleaner to group related things in a single part. Otherwise, you could have annotated reconcilers with just `reconciliationMaxIntervalTimeUnit` and that would look really weird. Currently, the only hint that these things are related are their names. It's better to make it explicit, in my opinion. It also avoids polluting the main configuration with things that will rarely be used. Embedding the unit in the String is a recipe for disaster, imo, and I think it's much cleaner to separate the configuration into a separate annotation.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: So just to see it clear. The reason to separate this into a new annotation because there are two related fields? 
I'm not sure that makes the effort, especially since we everything in one (for now mandatory) annotation what is a good thing. 
We would like to have this feature by default turned on, so what if that annotation is not present? Actually I guess should be turned off? It's easy to miss that it even exists.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: Agree that it nicer, but also has these side effects..so I lean little bit more to the current way. But if you insist we can change it. Just lets clarify what should be the behavior.

Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: Which side effects?

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: As mentioned, wanted to have this behavior turned on by default. So the developers should turn it off if they are absolutely sure that the operator works as intended (maybe not even that case). But if this is a separate annotation, then it needs to be put there explicitly, otherwise it would be quite contra intuitive to have the functionality turned on without the annotation.

Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: The annotation is only needed if you want to change the default values. We already do that in lots of other cases.

The only scenario where things are not clear is if you want to turn it off completely. Then again, that's not clear either with the current implementation.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: Where do we do that? What I mean is, that in spring and other frameworks  a functionality is turned on by adding an annotation, for me it sounds a little smelly that if we have an annotation for that but its not there and still turned on.

I created a constant for that to turn off, or any non negative value would do. It's quite common to do it this way.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: Maybe we should ask the users tbh. @andreaTP @lburgazzoli what do you think, should be this a separate annotation or not:
https://github.com/java-operator-sdk/java-operator-sdk/issues/848

File: docs/documentation/features.md (hunk 2 above)
Comment (csviri) [reply docs/documentation/features.md]: hehe, too late :)

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java (hunk 4 above)
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: The namespace control is done that way. Pretty much all the annotation fields have default values so we already turn things on without the annotation being present. This would be the exact same thing here: the child annotation would just allow people to change the default value.

Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: To be clear, I'm not saying we should have a completely separate annotation, just a composite child annotation for this.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: > To be clear, I'm not saying we should have a completely separate annotation, just a composite child annotation for this.

Ahhh, ok that is a completely different story! Sure that makes sense! Sorry....

Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ControllerConfiguration.java:119]: I just understood that I hadn't been clear enough with what I was saying with your last commit :)
Sorry! 😓

File: docs/documentation/features.md [hunk 5]
```diff
@@ -223,6 +221,28 @@ public class DeploymentReconciler
...
+
+In case informers are all in place and reconciler is implemented correctly, there is no need for additional triggers. 
+However, it's a [common practice](https://github.com/java-operator-sdk/java-operator-sdk/issues/848#issuecomment-1016419966) 
+to have a failsafe periodic trigger in place,
+just to make sure the resources are reconciled after certain time. This functionality is in place by default, there
+is quite high interval (currently 10 hours) while the reconciliation is triggered. See how to override this using 
+the standard annotation:
+
+```java
+@ControllerConfiguration(finalizerName = NO_FINALIZER, reconciliationMaxInterval = 2L,
```
Comment (metacosm) [inline docs/documentation/features.md:234]: This needs to be updated.

**Review by metacosm — APPROVED** [review]

**Review by metacosm — CHANGES_REQUESTED** [review]
LGTM apart the doc that needs to be updated. I might open a subsequent PR to mirror the annotation split in the configuration itself later.

**Review by metacosm — APPROVED** [review]
Cleaner, I like it! 👍🏼
